### PR TITLE
Server tests for Entropy

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Fixes https://github.com/aspnet/Coherence-Signed/issues/113

Please review @Tratcher

### CODE REVIEW

**Comment by victorhurdugaci** [comment]
This also adds support for Travis & AppVeyor 
cc @Eilon

File: build.cmd [hunk 1]
```diff
@@ -32,6 +32,8 @@ IF "%BUILDCMD_DNX_VERSION%"=="" (
 )
 IF "%SKIP_DNX_INSTALL%"=="" (
     CALL packages\KoreBuild\build\dnvm install %BUILDCMD_DNX_VERSION% -runtime CoreCLR -arch x86 -alias default
+    CALL packages\KoreBuild\build\dnvm install default -runtime CoreCLR -arch x64
+    CALL packages\KoreBuild\build\dnvm install default -runtime CLR -arch x64
     CALL packages\KoreBuild\build\dnvm install default -runtime CLR -arch x86 -alias default
```
Comment (Tratcher) [inline build.cmd]: Why is default set to both CoreCLR and CLR?

File: global.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
-    "projects": [
-        "src"
-    ]
+  "projects": [
+    "samples",
+    "src"
```
Comment (Tratcher) [inline global.json]: The src/Placeholder can be removed at this point.

File: samples/Builder.Filtering.Web/project.json [hunk 3]
```diff
@@ -1,14 +1,16 @@
 {
   "dependencies": {
     "Microsoft.AspNet.Http": "1.0.0-*",
-    "Microsoft.AspNet.Server.IIS" : "1.0.0-*",
-    "Microsoft.AspNet.Server.WebListener" : "1.0.0-*"
+    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
```
Comment (Tratcher) [inline samples/Builder.Filtering.Web/project.json]: Replace Microsoft.AspNet.Server.IIS with Microsoft.AspNet.IISPlatformHandler.

File: samples/Builder.Filtering.Web/project.json [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
+    "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
+    "Microsoft.AspNet.Server.WebListener": "1.0.0-*"
   },
   "frameworks": {
     "dnx451": { },
     "dnxcore50" : { }
   },
   "commands": {
```
Comment (Tratcher) [inline samples/Builder.Filtering.Web/project.json]: Change the commands to 
```
    "web": "Microsoft.AspNet.Server.Kestrel",
    "weblistener": "Microsoft.AspNet.Server.WebListener"
```
Compare to https://github.com/aspnet/Security/pull/597/files#diff-68483d081e22a22c9f6fa534b0a1abea

File: Entropy.sln [hunk 5]
```diff
@@ -100,6 +94,12 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "MusicStore.Spa", "samples\M
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Owin.ConsoleApp.HelloWorld", "samples\Owin.ConsoleApp.HelloWorld\Owin.ConsoleApp.HelloWorld.xproj", "{2CB93CF9-D432-4356-874F-3F897C2F291E}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "samples", "samples", "{8B204FB8-31A6-4559-AE7F-0412F504983C}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{C03BF7E1-ECDF-48D0-85CD-A454AA28D80C}"
+EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "ServerTests", "test\BuilderTests\ServerTests.xproj", "{42409264-0678-4C62-A8E2-702DC1177F39}"
```
Comment (Tratcher) [inline Entropy.sln]: Rename to FunctionalTests

File: test/ServerTests/ContainerTests/ContainerFallbackWebTests.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+using Microsoft.AspNet.Testing.xunit;
+using Xunit;
+
+namespace ServerTests.ContainerTests
+{
+    public class ContainerFallbackWebTests
+    {
+        private const string SiteName = "Container.Fallback.Web";
+
+        [ConditionalTheory]
```
Comment (Tratcher) [inline test/ServerTests/ContainerTests/ContainerFallbackWebTests.cs]: Are you going to have any tests for IIS Express + Kestrel?

File: test/ServerTests/ServerHelper.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+namespace ServerTests
+{
+    public static class ServerHelper
+    {
+        private static string GetPathToApplication(string applicationName)
+        {
+            return Path.GetFullPath(Path.Combine("..", "..", "samples", applicationName));
+        }
+
+        public static async Task ServerResponseMessageTest(string siteName, ServerType serverType, RuntimeFlavor runtimeFlavor, RuntimeArchitecture architecture, string applicationBaseUrl,
```
Comment (Tratcher) [inline test/ServerTests/ServerHelper.cs]: ServerResponseMessageTest -> RunSite

File: test/ServerTests/ServerHelper.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            var logger = new LoggerFactory()
+                            .AddConsole()
+                            .CreateLogger(string.Format("ResponseFormats:{0}:{1}:{2}", serverType, runtimeFlavor, architecture));
+
+            using (logger.BeginScope("ResponseFormatsTest"))
+            {
+                var deploymentParameters = new DeploymentParameters(GetPathToApplication(siteName), serverType, runtimeFlavor, architecture)
+                {
+                    ApplicationBaseUriHint = applicationBaseUrl,
+                    Command = serverType == ServerType.WebListener ? "web" : "kestrel",
```
Comment (Tratcher) [inline test/ServerTests/ServerHelper.cs]: Remember to update this when you change the commands.

File: test/ServerTests/Diagnostics/DiagnosticsStatusCodesMvcTests.cs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                architecture,
+                applicationBaseUrl,
+                async (httpClient, logger, token) =>
+                {
+                    var response = await RetryHelper.RetryRequest(async () =>
+                    {
+                        return await httpClient.GetAsync("/errors/" + (int)HttpStatusCode.PaymentRequired);
+                    }, logger, token);
+
+                    var responseText = await response.Content.ReadAsStringAsync();
```
Comment (Tratcher) [inline test/ServerTests/Diagnostics/DiagnosticsStatusCodesMvcTests.cs]: Try/Catch&Log the response text if the Assert fails. Otherwise it's hard to debug tests on the CI. See https://github.com/aspnet/IISIntegration/blob/dev/test/Microsoft.AspNet.IISPlatformHandler.FunctionalTests/NtlmAuthentationTest.cs#L116-L121

File: test/ServerTests/ServerHelper.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+                    SiteName = "HttpTestSite",
+                };
+
+                using (var deployer = ApplicationDeployerFactory.Create(deploymentParameters, logger))
+                {
+                    var deploymentResult = deployer.Deploy();
+                    var httpClientHandler = new HttpClientHandler();
+                    var httpClient = new HttpClient(httpClientHandler) { BaseAddress = new Uri(deploymentResult.ApplicationBaseUri) };
```
Comment (Tratcher) [inline test/ServerTests/ServerHelper.cs]: Set the timeout to something like 5 seconds: https://github.com/aspnet/IISIntegration/blob/dev/test/Microsoft.AspNet.IISPlatformHandler.FunctionalTests/NtlmAuthentationTest.cs#L49  Otherwise it may wait 100 seconds.

File: test/ServerTests/Diagnostics/DiagnosticsStatusCodesMvcTests.cs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                server,
+                runtimeFlavor,
+                architecture,
+                applicationBaseUrl,
+                async (httpClient, logger, token) =>
+                {
+                    var response = await RetryHelper.RetryRequest(async () =>
+                    {
+                        return await httpClient.GetAsync("/errors/" + (int)HttpStatusCode.PaymentRequired);
+                    }, logger, token);
```
Comment (Tratcher) [inline test/ServerTests/Diagnostics/DiagnosticsStatusCodesMvcTests.cs]: Set the retry count to 15 or 30 (default 60). https://github.com/aspnet/IISIntegration/blob/dev/test/Microsoft.AspNet.IISPlatformHandler.FunctionalTests/NtlmAuthentationTest.cs#L56

File: test/ServerTests/ContainerTests/ContainerFallbackWebTests.cs (hunk 6 above)
Comment (victorhurdugaci) [inline test/ServerTests/ContainerTests/ContainerFallbackWebTests.cs]: Nope

**Comment by victorhurdugaci** [comment]
Updated

File: build.cmd (hunk 1 above)
Comment (victorhurdugaci) [inline build.cmd]: The second one overrides the first

File: samples/Content.Upload.Files/project.json [hunk 12]
```diff
@@ -10,7 +10,7 @@
     "**.vspscc"
   ],
   "dependencies": {
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*"
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*"
```
Comment (Tratcher) [inline samples/Content.Upload.Files/project.json]: This one needs Kestrel & WebListener & some commands

File: samples/Content.Upload.Form/project.json [hunk 13]
```diff
@@ -10,7 +10,7 @@
     "**.vspscc"
   ],
   "dependencies": {
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*"
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*"
```
Comment (Tratcher) [inline samples/Content.Upload.Form/project.json]: This one needs Kestrel & WebListener & some commands

File: samples/Content.Upload.Multipart/project.json [hunk 14]
```diff
@@ -10,7 +10,7 @@
     "**.vspscc"
   ],
   "dependencies": {
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*",
```
Comment (Tratcher) [inline samples/Content.Upload.Multipart/project.json]: This one needs Kestrel & WebListener & some commands

File: samples/Logging.Elm/Properties/launchSettings.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "Hosting:Environment": "Development"
+      }
+    },
+    "web": {
+      "commandName": "web",
+      "environmentVariables": {
+        "Hosting:Environment": "Development"
```
Comment (Tratcher) [inline samples/Logging.Elm/Properties/launchSettings.json]: Hosting:Environment isn't valid anymore, it should be ASPNET_ENVIRONMENT. Are you using RC1 tooling?

File: samples/Content.Upload.Form/project.json (hunk 13 above)
Comment (victorhurdugaci) [inline samples/Content.Upload.Form/project.json]: Lol. Whoever wrote these apparently never ran them...

File: samples/Mvc.CustomRoutingConvention/project.json [hunk 16]
```diff
@@ -1,7 +1,7 @@
 {
   "dependencies": {
     "Microsoft.AspNet.Mvc": "6.0.0-*",
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*"
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*"
   },
   "commands": {
     "web": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.WebListener --server.urls http://localhost:5000"
```
Comment (Tratcher) [inline samples/Mvc.CustomRoutingConvention/project.json]: Update the commands and add WebListener and Kestrel dependencies.

File: samples/Logging.Elm/Properties/launchSettings.json (hunk 15 above)
Comment (victorhurdugaci) [inline samples/Logging.Elm/Properties/launchSettings.json]: This file shouldn't be here.

File: samples/Mvc.Modules/project.json [hunk 17]
```diff
@@ -1,7 +1,7 @@
 {
   "dependencies": {
     "Microsoft.AspNet.Mvc": "6.0.0-*",
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*",
     "Microsoft.AspNet.Server.WebListener": "1.0.0-*",
```
Comment (Tratcher) [inline samples/Mvc.Modules/project.json]: Needs Kestrel & command

File: samples/Owin.IAppBuilderBridge/project.json [hunk 18]
```diff
@@ -15,7 +15,7 @@
   "dependencies": {
     "Microsoft.AspNet.Owin": "1.0.0-*",
     "Microsoft.AspNet.DataProtection": "1.0.0-*",
-    "Microsoft.AspNet.Server.IIS": "1.0.0-*",
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*",
     "Microsoft.Extensions.DependencyInjection": "1.0.0-*",
```
Comment (Tratcher) [inline samples/Owin.IAppBuilderBridge/project.json]: Needs Kestrel & command

File: samples/Mvc.CustomRoutingConvention/project.json (hunk 16 above)
Comment (victorhurdugaci) [inline samples/Mvc.CustomRoutingConvention/project.json]: I've only updated the projects that I cover with tests

File: global.json (hunk 2 above)
Comment (Tratcher) [inline global.json]: You removed Placeholder, but left src?

File: samples/Owin.IAppBuilderBridge/project.json (hunk 18 above)
Comment (victorhurdugaci) [inline samples/Owin.IAppBuilderBridge/project.json]: Not fixing projects that I don't cover with tests

File: test/FunctionalTests/project.json [hunk 19]
```diff
@@ -0,0 +1,24 @@
+﻿{
+  "compilationOptions": {
+    "warningsAsErrors": true
+  },
+  "commands": {
+    "test": "xunit.runner.aspnet"
+  },
+  "dependencies": {
+    "Microsoft.AspNet.IISPlatformHandler": "1.0.0-*",
```
Comment (Tratcher) [inline test/FunctionalTests/project.json]: I don't think you need IISPlatformHandler in the test launcher, just in the samples.

File: samples/Mvc.Modules/project.json (hunk 17 above)
Comment (victorhurdugaci) [inline samples/Mvc.Modules/project.json]: Not fixing projects that I don't cover with tests. Open a bug for it

File: test/FunctionalTests/TestServices.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public static async Task RunSiteTest(string siteName, ServerType serverType, RuntimeFlavor runtimeFlavor, RuntimeArchitecture architecture, string applicationBaseUrl,
+            Func<HttpClient, ILogger, CancellationToken, Task> validator)
+        {
+            var logger = new LoggerFactory()
+                            .AddConsole()
+                            .CreateLogger(string.Format("ResponseFormats:{0}:{1}:{2}", serverType, runtimeFlavor, architecture));
+
+            using (logger.BeginScope("ResponseFormatsTest"))
```
Comment (Tratcher) [inline test/FunctionalTests/TestServices.cs]: Logging names are out of date.

File: samples/Owin.IAppBuilderBridge/project.json (hunk 18 above)
Comment (Tratcher) [inline samples/Owin.IAppBuilderBridge/project.json]: Then why modify the project.json at all? Half fixing it just hides the problem.

Comment (victorhurdugaci) [inline samples/Owin.IAppBuilderBridge/project.json]: Removed unrelated changes

**Comment by victorhurdugaci** [comment]
Updated

File: test/FunctionalTests/TestServices.cs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+                throw;
+            }
+        }
+
+        public static async Task RunSiteTest(string siteName, ServerType serverType, RuntimeFlavor runtimeFlavor, RuntimeArchitecture architecture, string applicationBaseUrl,
+            Func<HttpClient, ILogger, CancellationToken, Task> validator)
+        {
+            var logger = new LoggerFactory()
+                            .AddConsole()
+                            .CreateLogger(string.Format("ResponseFormats:{0}:{1}:{2}", serverType, runtimeFlavor, architecture));
```
Comment (Tratcher) [inline test/FunctionalTests/TestServices.cs]: This logger text too.

File: test/FunctionalTests/Properties/AssemblyInfo.cs [hunk 22]
```diff
@@ -0,0 +1,23 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("BuilderTests")]
```
Comment (Tratcher) [inline test/FunctionalTests/Properties/AssemblyInfo.cs]: Delete all of this.

File: test/FunctionalTests/TestServices.cs (hunk 21 above)
Comment (victorhurdugaci) [inline test/FunctionalTests/TestServices.cs]: whoops

**Comment by Tratcher** [comment]
:shipit:
